### PR TITLE
Add C++17 std::filesystem Extension

### PR DESCRIPTION
**Pull request by time-killer-games**
TODO:
- get BSD pr merged so the Makefile won't throw an error (UNIX_BASED is defined in my BSD pr)
- replace try/catch exceptions with if checks to prevent segfaulting without making bigger/slower exe
- replace filename_normalize logic written in Win32/POSIX with C++17 std::filesystem::canonical
- replace create_directory with create_directories (this one is a one line fix so it'll be a quickie)
- replace overriding logic using macros to actually replace the current code so no override required
- add special protections on directory_destroy() since Josh is a big stinky cheese butt poopy head!
- add file_find_first(), file_find_next(), and file_find_close() (file_find_close() is a redundant stub tho)
- anything Josh, Robert, or fundies can add to this that I am missing. Please suggest your ideas.

Note i want this extension to be self-contained, meaning that all code can be used as-is in other projects unrelated to enigma without relying on code found in the rest of ENIGMA's repository.

Right now, the only things I duplicated into the extension for self-containment, are a few small string utility functions, which takes up about maybe 25 lines of code at most, and won't need maintenance practically ever, due to the how it was written, and how reliable the code itself is. I doubt it'll change.

### CODE REVIEW

File: ENIGMAsystem/SHELL/Universal_System/Extensions/FileManip/fileman.cpp [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+#include "fileman.h"
+
+using std::string;
+
+namespace enigma_user {
+
+  string get_working_directory_ns() {
+    return fileman::get_working_directory_ns();
+  }
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/FileManip/fileman.cpp:35]: Here/below:
```suggestion
using ::fileman::get_working_directory_ns;
```

File: ENIGMAsystem/SHELL/Universal_System/Extensions/FileManip/Universal/fileman.cpp [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+    return fname.substr(fp);
+  }
+
+} // namespace strings
+
+namespace fileman {
+
+  string get_working_directory_ns() {
+    string result = "";
+    try {
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/FileManip/Universal/fileman.cpp:87]: This is likely to miss other exceptions; you should favor the noexcept versions.

File: ENIGMAsystem/SHELL/Universal_System/Extensions/FileManip/fileman.h [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+  bool directory_rename_ns(std::string oldname, std::string newname);
+
+  bool directory_copy_ns(std::string dname, std::string newname);
+
+  std::string directory_contents_ns(std::string dname);
+
+  std::string directory_contents_ext_ns(std::string dname, std::string pattern, bool includedirs);
+
+  std::string environment_get_variable_ns(std::string name);
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/FileManip/fileman.h:131]: Most of this doesn't belong in a file manipulation extension, but especially these two.

File: ENIGMAsystem/SHELL/Universal_System/Extensions/FileManip/POSIX/fileman.cpp [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ SOFTWARE.
+ 
+*/
+
+#include "../fileman.h"
+#include <unistd.h>
+#include <cstdlib>
```
Comment (RobertBColton) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/FileManip/POSIX/fileman.cpp:29]: C++ headers should be included before C headers, because they often include C headers themselves.

Comment (RobertBColton) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/FileManip/POSIX/fileman.cpp:29]: C++ headers should be included before C headers, because they often include C headers themselves. Switch these two lines.

Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/FileManip/POSIX/fileman.cpp:29]: unistd.h is not a part of the C standard, it is a posix header.

Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/FileManip/POSIX/fileman.cpp:29]: Well the general rule is that C++ headers are included before C headers. If you start doing this now, other people get confused and will end up including more C headers there. That's why it's important to be consistent across all source files about the include order.

Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/FileManip/POSIX/fileman.cpp:29]: Also, it doesn't matter if `unistd.h` is part of the C standard, it's still a C header indicated by the `.h` extension.

Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/FileManip/POSIX/fileman.cpp:29]: Thank you for the clarification. I just wanted to double check. Does this also apply to Windows headers like windows.h and shlobj.h? I guess it's safe to assume this isnt the only file you want changed, but pretty much all of them lol will do

Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/FileManip/POSIX/fileman.cpp:29]: Yeah, the Windows header is a C header. Also, yeah, it's easier to do it now rather than later.

File: ENIGMAsystem/SHELL/Universal_System/Extensions/FileManip/include.h [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ SOFTWARE.
+ 
+*/
+
+#define OVERRIDE_SANDBOX
```
Comment (RobertBColton) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/FileManip/include.h:27]: What the hell is sandbox?

File: ENIGMAsystem/SHELL/Universal_System/Extensions/FileManip/fileman.h [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+*/
+
+#include <string>
+#include <vector>
+#include <cstddef>
+
+namespace strings {
+
+  std::string string_replace_all(std::string str, std::string substr, std::string nstr);
+  std::vector<std::string> string_split(std::string str, char delimiter);
```
Comment (RobertBColton) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/FileManip/fileman.h:34]: How are these file related? If they aren't then you should use the ones we already have so we stop duplicating them.

Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/FileManip/fileman.h:34]: I already had explained this several months ago. I'm done repeating myself to you and Josh about literally everything.

File: ENIGMAsystem/SHELL/Universal_System/Extensions/FileManip/include.h (hunk 5 above)
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/FileManip/include.h:27]: Are you mad at me or something? What has gotten with you today? The drama never ends. I get 3 reviews in two minutes all 3 of them seem to be intentionally trying to get on my nerves rather than do anything productive. 

That said, the current functions are going to be sandboxed, as if you didnt know that, we had a large discussion about sandboxing the file system and this overrides that behavior by turning on the extension checkbox. Sheesh.
